src/main/java/org/breedinginsight/services/TraitService.java

```diff
@@ -59,4 +93,158 @@
+                            .methodName(trait.getMethod().getMethodName())
+                            .programOntologyId(programOntology.getId())
+                            .createdBy(actingUser.getId())
+                            .updatedBy(actingUser.getId())
+                            .build();
+                    methodDAO.insert(jooqMethod);
+                    trait.getMethod().setId(jooqMethod.getId());
+
+                    // Create scale
+                    ScaleEntity jooqScale = ScaleEntity.builder()
+                            .scaleName(trait.getScale().getScaleName())
+                            .dataType(trait.getScale().getDataType())
+                            .programOntologyId(programOntology.getId())
+                            .createdBy(actingUser.getId())
+                            .updatedBy(actingUser.getId())
+                            .build();
+                    scaleDAO.insert(jooqScale);
+                    trait.getScale().setId(jooqScale.getId());
+
+                    // Create trait
+                    TraitEntity jooqTrait = TraitEntity.builder()
+                            .traitName(trait.getTraitName())
+                            .abbreviations(trait.getAbbreviations())
+                            .programOntologyId(programOntology.getId())
+                            .programObservationLevelId(trait.getProgramObservationLevel().getId())
+                            .methodId(jooqMethod.getId())
+                            .scaleId(jooqScale.getId())
+                            .createdBy(actingUser.getId())
+                            .updatedBy(actingUser.getId())
+                            .build();
+                    traitDAO.insert(jooqTrait);
+                    trait.setId(jooqTrait.getId());
+                }
+
+                return traitDAO.createTraitsBrAPI(traits, user, program);
+            });
+        }
+
+        return createdTraits;
+    }
+
+    public void assignTraitsProgramObservationLevel(List<Trait> traits, UUID programId, TraitValidatorErrorInterface traitValidatorError) throws ValidatorException {
+
+        ValidationErrors validationErrors = new ValidationErrors();
+
+        // Get our program observation levels
+        List<ProgramObservationLevel> programLevels = programObservationLevelService.getByProgramId(programId);
+        for (int i = 0; i < traits.size(); i++) {
+            Trait trait = traits.get(i);
+            if (trait.getProgramObservationLevel() != null){
+                List<ProgramObservationLevel> matchingLevels = programLevels.stream()
+                        .filter(programObservationLevel -> programObservationLevel.getName().equals(trait.getProgramObservationLevel().getName()))
+                        .collect(Collectors.toList());
+                if (matchingLevels.size() == 0) {
+                    ValidationError validationError = traitValidatorError.getTraitLevelDoesNotExist();
```

Comment:
Talked about adding list of available levels to error message so user can know what is a valid level.

Reply:
Changed and pushed